### PR TITLE
Changed grpc_error_get|set_str to use std string instead of slice

### PR DESCRIPTION
**Pull request by veblush**
Currently `grpc_error_get_str` and `grpc_error_set_str`use a `slice` type value for a string. This isn't compatible with `absl::Status` since `grpc_error_get_str` returns a slice without increasing a reference counter, which is not possible to implement with `absl::Status`. Therefore, `std::string` and `absl::string_view` are used for these functions as a type of value. This may introduce a bit of overhead in case of `grpc_error` (not `absl::Status`) but this isn't on a critical path so it'd be acceptable.

---
This change is [Reviewable](https://reviewable.io/reviews/grpc/grpc/27466)

### CODE REVIEW

**Review by veblush — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/grpc/grpc/27466)* status: all files reviewed, 11 unresolved discussions (waiting on @markdroth)

___
*[src/core/lib/iomgr/error.h, line 367 at r1](https://reviewable.io/reviews/grpc/grpc/27466#-MkbtRPv-K1uyeWwuR7H:-MkdhLky9TWapC4pYcvY:b-896fix) ([raw file](https://github.com/grpc/grpc/blob/e2d35c48ee4bf764cfedef56a8c90bdd9ee1c24b/src/core/lib/iomgr/error.h#L367)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

This comment can go away now.
</blockquote></details>

Done.
___
*[src/core/lib/iomgr/error.h, line 373 at r1](https://reviewable.io/reviews/grpc/grpc/27466#-MkbtSyz-P6ftNprq87n:-MkdhPnn-JUhK0scJAin:b-896fix) ([raw file](https://github.com/grpc/grpc/blob/e2d35c48ee4bf764cfedef56a8c90bdd9ee1c24b/src/core/lib/iomgr/error.h#L373)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

This comment can go away now.
</blockquote></details>

Done.
___
*[src/core/lib/iomgr/error.h, line 375 at r1](https://reviewable.io/reviews/grpc/grpc/27466#-MkbtUJG6PalL-uzFqij:-MkdgWga9M-Zfs7mamIM:b-zeum9b) ([raw file](https://github.com/grpc/grpc/blob/e2d35c48ee4bf764cfedef56a8c90bdd9ee1c24b/src/core/lib/iomgr/error.h#L375)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

Any reason not to use `absl::string_view` here?  Seems like that would avoid an unnecessary copy, since the caller is already holding a ref to the error that contains the string.

Reading further, I see that the issue is that `StatusGetStr()` returns a `std::string`, since the underlying `Cord` might be non-contiguous.  Would it make sense to change `StatusGetStr()` such that it returns an `absl::string_view` but takes as a parameter a `std::string*` pointing to a caller-owned buffer to use in the case where the underlying string is non-contiguous?  That would allow us to avoid a copy in the case where it is contiguous, which seems likely to be the common case for our uses.  (We use that approach in [`grpc_metadata_batch_get_value()`](https://github.com/grpc/grpc/blob/d30404868f2bf7725f5dd36245e1149a09349bfd/src/core/lib/transport/metadata_batch.h#L611), and it's proven to work fairly well.)
</blockquote></details>

I considered using `absl::string_view` first but it's generally discouraged unless it's absolutely desired mainly it makes it complicated. (Callers should be aware of the life-time of referenced string or cord). I thought `grpc_error_get_str` wasn't used on a performance critical path so I decide to stick with a simple `std::string`. 

Do you think it's really important to avoid one more copy here?
___
*[src/core/lib/iomgr/error.cc, line 601 at r1](https://reviewable.io/reviews/grpc/grpc/27466#-Mkbtrvh3gHwqae0yEAr:-MkdhuUj4lS-3W_fGM0c:b-896fix) ([raw file](https://github.com/grpc/grpc/blob/e2d35c48ee4bf764cfedef56a8c90bdd9ee1c24b/src/core/lib/iomgr/error.cc#L601)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

In this particular case, we don't already have a `std::string`, so constructing one and then using `grpc_slice_from_cpp_string()` is two separate allocations.  I think it would make more sense to use `grpc_slice_from_copied_buffer()` instead here, since that would be just one allocation.
</blockquote></details>

Done.
___
*[src/core/lib/iomgr/tcp_client_posix.cc, line 222 at r1](https://reviewable.io/reviews/grpc/grpc/27466#-MkbxkJV0D-l_3rVcjpW:-Mkdm7S0D-f8UezXPOZV:bxo41fs) ([raw file](https://github.com/grpc/grpc/blob/e2d35c48ee4bf764cfedef56a8c90bdd9ee1c24b/src/core/lib/iomgr/tcp_client_posix.cc#L222)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

This comment can go away now.
</blockquote></details>

I don't understand what the original comments mean?
___
*[src/core/lib/surface/call.cc, line 747 at r1](https://reviewable.io/reviews/grpc/grpc/27466#-MkbyFuf55vcZFkaUZAc:-MkdiMe5AoTgkryr0Q3Q:b-896fix) ([raw file](https://github.com/grpc/grpc/blob/e2d35c48ee4bf764cfedef56a8c90bdd9ee1c24b/src/core/lib/surface/call.cc#L747)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

This comment can go away now.
</blockquote></details>

Done.
___
*[src/core/lib/transport/error_utils.cc, line 72 at r1](https://reviewable.io/reviews/grpc/grpc/27466#-Mkbushb3_QhfLlRdT9C:-MkdinRB8_R1FucZg8ia:b9secji) ([raw file](https://github.com/grpc/grpc/blob/e2d35c48ee4bf764cfedef56a8c90bdd9ee1c24b/src/core/lib/transport/error_utils.cc#L72)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

s/slice/string/

Same thing on the next line.
</blockquote></details>

Instead, I chose `message` here because the name of the variable is `message`
___
*[test/core/iomgr/error_test.cc, line 61 at r1](https://reviewable.io/reviews/grpc/grpc/27466#-Mkbz5yu1D6R2qvnpfIv:-MkdkRtk5tBN635REUX7:bff5z24) ([raw file](https://github.com/grpc/grpc/blob/e2d35c48ee4bf764cfedef56a8c90bdd9ee1c24b/test/core/iomgr/error_test.cc#L61)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

I don't think this actually works if you're not using the gtest framework here, because without that framework it has no way to cause the test binary to exit with a non-zero status.
</blockquote></details>

Good catch! I'm changing it to use gtest. 
___
*[test/core/security/aws_request_signer_test.cc, line 249 at r1](https://reviewable.io/reviews/grpc/grpc/27466#-Mkbyqlb7PuXVV3LICVq:-MkdkscjA6YXXMeIKlYN:b-896fix) ([raw file](https://github.com/grpc/grpc/blob/e2d35c48ee4bf764cfedef56a8c90bdd9ee1c24b/test/core/security/aws_request_signer_test.cc#L249)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

This can just be a `const char*` to avoid an unnecessary allocation.

Or better yet, just inline this into the `EXPECT_EQ()` call below.
</blockquote></details>

Done.
___
*[test/core/security/aws_request_signer_test.cc, line 263 at r1](https://reviewable.io/reviews/grpc/grpc/27466#-MkbytNED2ulaFHzIaGx:-MkdksxqEz7qU2gjWD2X:b-896fix) ([raw file](https://github.com/grpc/grpc/blob/e2d35c48ee4bf764cfedef56a8c90bdd9ee1c24b/test/core/security/aws_request_signer_test.cc#L263)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

Same here.
</blockquote></details>

Done.
___
*[test/core/security/credentials_test.cc, line 3225 at r1](https://reviewable.io/reviews/grpc/grpc/27466#-MkbzLYfDROzTzZUuAv9:-Mkdm1mvC9Zwk8kOcFeD:b-896fix) ([raw file](https://github.com/grpc/grpc/blob/e2d35c48ee4bf764cfedef56a8c90bdd9ee1c24b/test/core/security/credentials_test.cc#L3225)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

Same comment as elsewhere: I don't think it's okay to use gmock expectations without using gtest as the framework.
</blockquote></details>

Done.

**Review by markdroth — CHANGES_REQUESTED** [review]
> Reviewed 7 of 7 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/grpc/grpc/27466)* status: all files reviewed, 3 unresolved discussions (waiting on @veblush)

___
*[src/core/lib/iomgr/error.h, line 375 at r1](https://reviewable.io/reviews/grpc/grpc/27466#-MkbtUJG6PalL-uzFqij:-MkhLgyy32-wbEtZkRYJ:b5yyrju) ([raw file](https://github.com/grpc/grpc/blob/e2d35c48ee4bf764cfedef56a8c90bdd9ee1c24b/src/core/lib/iomgr/error.h#L375)):*
<details><summary><i>Previously, veblush (Esun Kim) wrote…</i></summary><blockquote>

I considered using `absl::string_view` first but it's generally discouraged unless it's absolutely desired mainly it makes it complicated. (Callers should be aware of the life-time of referenced string or cord). I thought `grpc_error_get_str` wasn't used on a performance critical path so I decide to stick with a simple `std::string`. 

Do you think it's really important to avoid one more copy here?
</blockquote></details>

I think the interface isn't that much more complicated to support `absl::string_view`, so it seems useful.  But I'll leave it up to you.
___
*[src/core/lib/iomgr/tcp_client_posix.cc, line 222 at r1](https://reviewable.io/reviews/grpc/grpc/27466#-MkbxkJV0D-l_3rVcjpW:-MkhKLAQ4UwQ0HkKd_GJ:b845nle) ([raw file](https://github.com/grpc/grpc/blob/e2d35c48ee4bf764cfedef56a8c90bdd9ee1c24b/src/core/lib/iomgr/tcp_client_posix.cc#L222)):*
<details><summary><i>Previously, veblush (Esun Kim) wrote…</i></summary><blockquote>

I don't understand what the original comments mean?
</blockquote></details>

The original code was getting a slice from `ac->addr_str` before it released the lock.  You're now looking at that field *after* releasing the lock (which I assume you've verified is safe), so the comment no longer applies.
___
*[test/core/iomgr/error_test.cc, line 33 at r2](https://reviewable.io/reviews/grpc/grpc/27466#-MkhKhLA-DUIICzFywqR:-MkhKhLB9ERQ727N6hIB:b-gpiqv7) ([raw file](https://github.com/grpc/grpc/blob/21641030c6213f1c18a43ff98b2ee3acc46197e8/test/core/iomgr/error_test.cc#L33)):*
> ```cpp
> TEST(ErrorTest, SetGetInt) {
>   grpc_error_handle error = GRPC_ERROR_CREATE_FROM_STATIC_STRING("Test");
>   GPR_ASSERT(error != GRPC_ERROR_NONE);
> ```

Since you're converting this to gtest, can you also replace all of the `GPR_ASSERT()` calls with the appropriate `EXPECT_*` (or, rarely, `ASSERT_*`) gtest macros?
___
*[test/core/security/credentials_test.cc, line 315 at r2](https://reviewable.io/reviews/grpc/grpc/27466#-MkhLEW-4gwaAcm-T5ad:-MkhLEW-4gwaAcm-T5ae:b-kn58n0) ([raw file](https://github.com/grpc/grpc/blob/21641030c6213f1c18a43ff98b2ee3acc46197e8/test/core/security/credentials_test.cc#L315)):*
> ```cpp
>   grpc_credentials_mdelem_array md_array;
>   md_array = {};
>   GPR_ASSERT(md_array.md == nullptr);
> ```

Same comment as in the other test: Can you also change all of the `GPR_ASSERT()` calls to the appropriate `EXPECT_*` macro?

Alternatively, just leave a TODO at the top of the file to finish converting these tests to idiomatic C++.  There's probably a lot more restructing to be done here.

**Review by veblush — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/grpc/grpc/27466)* status: all files reviewed, 3 unresolved discussions (waiting on @markdroth)

___
*[src/core/lib/iomgr/error.h, line 375 at r1](https://reviewable.io/reviews/grpc/grpc/27466#-MkbtUJG6PalL-uzFqij:-Mkhae_Y0Pz5bKMJxrKZ:bo4da40) ([raw file](https://github.com/grpc/grpc/blob/e2d35c48ee4bf764cfedef56a8c90bdd9ee1c24b/src/core/lib/iomgr/error.h#L375)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

I think the interface isn't that much more complicated to support `absl::string_view`, so it seems useful.  But I'll leave it up to you.
</blockquote></details>

I like to keep this way to make the API easy to use but I keep an eye on performance change once this gets merged.
___
*[src/core/lib/iomgr/tcp_client_posix.cc, line 222 at r1](https://reviewable.io/reviews/grpc/grpc/27466#-MkbxkJV0D-l_3rVcjpW:-MkhbG_lCFQ3FrDatyPf:bx7mrqj) ([raw file](https://github.com/grpc/grpc/blob/e2d35c48ee4bf764cfedef56a8c90bdd9ee1c24b/src/core/lib/iomgr/tcp_client_posix.cc#L222)):*
> ```
> Create a copy of the data from "ac" to be accessed after the unlock, as
> ```
Thanks! Done.
___
*[test/core/iomgr/error_test.cc, line 33 at r2](https://reviewable.io/reviews/grpc/grpc/27466#-MkhKhLA-DUIICzFywqR:-Mkhi8Wq-ypPb_uFCGtJ:b-896fix) ([raw file](https://github.com/grpc/grpc/blob/21641030c6213f1c18a43ff98b2ee3acc46197e8/test/core/iomgr/error_test.cc#L33)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

Since you're converting this to gtest, can you also replace all of the `GPR_ASSERT()` calls with the appropriate `EXPECT_*` (or, rarely, `ASSERT_*`) gtest macros?
</blockquote></details>

Done.
___
*[test/core/security/credentials_test.cc, line 315 at r2](https://reviewable.io/reviews/grpc/grpc/27466#-MkhLEW-4gwaAcm-T5ad:-Mkhi9019cNGODoHwRgJ:b-896fix) ([raw file](https://github.com/grpc/grpc/blob/21641030c6213f1c18a43ff98b2ee3acc46197e8/test/core/security/credentials_test.cc#L315)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

Same comment as in the other test: Can you also change all of the `GPR_ASSERT()` calls to the appropriate `EXPECT_*` macro?

Alternatively, just leave a TODO at the top of the file to finish converting these tests to idiomatic C++.  There's probably a lot more restructing to be done here.
</blockquote></details>

Done.

**Review by markdroth — APPROVED** [review]
Looks great!

> Reviewed 3 of 3 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/grpc/grpc/27466#-:-MkhjcnL4OKG1qaDkOGE:b9bkkp1)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @veblush)

**Comment by veblush** [comment]
Hmm it looks like cmake tests on mac has an issue ([log](https://source.cloud.google.com/results/invocations/853921dd-8d61-455b-b681-5668a3bb6698/targets/github%2Fgrpc%2Frun_tests%2Fcpp_macos_dbg_native/tests)).

```
[ RUN      ] CredentialsTest.google_default_creds_external_account_credentials
I0928 17:37:17.688600000 4388369856 check_gcp_environment_no_op.cc:28] ALTS: Platforms other than Linux and Windows are not supported
E0928 17:37:17.699525000 4388369856 google_default_credentials.cc:429] Could not create google default credentials: {"created":"@1632875837.688407000","description":"Invalid external account credentials format.","file":"/Volumes/BuildData/tmpfs/src/github/grpc/workspace_c++_macos_dbg_native/src/core/lib/security/credentials/google_default/google_default_credentials.cc","file_line":327,"referenced_errors":[{"created":"@1632875837.688592000","description":"Failed to load file","file":"/Volumes/BuildData/tmpfs/src/github/grpc/workspace_c++_macos_dbg_native/src/core/lib/iomgr/load_file.cc","file_line":72,"filename":"/Users/kbuilder/.config/gcloud/application_default_credentials.json","referenced_errors":[{"created":"@1632875837.688583000","description":"No such file or directory","errno":2,"file":"/Volumes/BuildData/tmpfs/src/github/grpc/workspace_c++_macos_dbg_native/src/core/lib/iomgr/load_file.cc","file_line":45,"os_error":"No such file or directory","syscall":"fopen"}]}]}
*** SIGSEGV received at time=1632875837 ***
PC: @        0x1021fcf8c  (unknown)  grpc_core::RefCountedPtr<>::get()
```

**Comment by veblush** [comment]
MacOS happens to hit the issue of running gtest tests in parallel which will be addressed by https://github.com/grpc/grpc/pull/27524.

**Comment by veblush** [comment]
This PR gtest-ified credentials_test but it has a weird runtime error when running by cmake-test on MacOS. (This only happens when running in parallel) This isn't reproducible on my mac machine so I decided to undo that part in this PR to unblock this since it's not directly relevant to the original motivation of this PR.

**Review by markdroth — APPROVED** [review]
That's unfortunate.  But I agree, it's not strictly required for this PR.

Can you point me to the error message?  I'm curious as to what the problem was.

> Reviewed 17 of 17 files at r4, 5 of 5 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/grpc/grpc/27466#-:-Mkx4f9H7OFv6mcY6IpO:bk9hpqj)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @veblush)

**Comment by veblush** [comment]
@markdroth https://github.com/grpc/grpc/pull/27466#issuecomment-930294336 is the one

**Pull request merged**
